In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv("details.csv")
df.head(3)

comment_id  season  match_id match_name home_team away_team  \
0         110  2023.0   1359475   GT v CSK        GT       CSK   
1         120  2023.0   1359475   GT v CSK        GT       CSK   
2         130  2023.0   1359475   GT v CSK        GT       CSK   

  current_innings  innings_id  over  ball  ...  wicket_id wkt_batsman_name  \
0             CSK           1     1     1  ...        NaN              NaN   
1             CSK           1     1     2  ...        NaN              NaN   
2             CSK           1     1     3  ...        NaN              NaN   

   wkt_bowler_name  wkt_batsman_runs  wkt_batsman_balls  wkt_text  \
0              NaN               NaN                NaN       NaN   
1              NaN               NaN                NaN       NaN   
2              NaN               NaN                NaN       NaN   

  isRetiredHurt                                               text  \
0         False  nice and full, angling and perhaps swinging in...   
1         False  Conway shuffles across off, and shimmies out a...   
2         False  good length, angling in at off and then straig...   

                                             preText  postText  
0  <p><strong>7.30pm</strong> Gaikwad and Conway ...       NaN  
1                                                NaN       NaN  
2                                                NaN       NaN  

[3 rows x 45 columns]

##### generally there are two methods to store data , 
##### 1] long data : no. of rows is extremely high than number of columns (such data we recieve from databases ) and
##### 2] wide data : no. of columns is extremely high than number of rows ( such data we recieve from datalakes , datawarehouses,webapis,webscraping)

In [3]:
df.shape

(17612, 45)

##### this shape indicates no.of rows=17612 , no. of columns=45 , this dataset is long data form 

#### melt method converts any dataframe to its maximum extent to become long data form

In [18]:
df.melt()

variable                                              value
0       comment_id                                                110
1       comment_id                                                120
2       comment_id                                                130
3       comment_id                                                140
4       comment_id                                                150
...            ...                                                ...
792535    postText                                                NaN
792536    postText                                                NaN
792537    postText                                                NaN
792538    postText                                                NaN
792539    postText  <p><b>3.01am</b> Dhoni collects the IPL trophy...

[792540 rows x 2 columns]

##### pivot_table method converts any dataset to maximum extent of wide data form , but moreover it is advanced version of groupby with more functionality , and acts as shorter method instead of writing multiple groupby statements , hence used widely

In [11]:
df.pivot_table(columns=['ball'],aggfunc='mean')

C:\Users\Admin\AppData\Local\Temp\ipykernel_5740\2503918822.py:1: FutureWarning: pivot_table dropped a column because it failed to aggregate. This behavior is deprecated and will raise in a future version of pandas. Select only the columns that can be aggregated.
  df.pivot_table(columns=['ball'],aggfunc='mean')


ball                          1             2             3             4  \
batsman1_balls     1.431922e+01  1.407407e+01  1.400271e+01  1.380054e+01   
batsman1_id        6.131202e+05  6.119854e+05  6.152186e+05  6.087029e+05   
batsman1_runs      2.019019e+01  2.041987e+01  2.027652e+01  2.000919e+01   
batsman2_balls     1.246808e+01  1.306970e+01  1.327652e+01  1.348502e+01   
batsman2_id        6.073352e+05  6.129574e+05  6.154794e+05  6.128023e+05   
batsman2_runs      1.790894e+01  1.831279e+01  1.868079e+01  1.904187e+01   
bowler1_id         6.966580e+05  6.956873e+05  6.960512e+05  6.949677e+05   
bowler1_maidens    1.108871e-02  1.043771e-02  1.016605e-02  1.055140e-02   
bowler1_overs      1.367036e+00  1.463401e+00  1.559031e+00  1.647754e+00   
bowler1_runs       1.150134e+01  1.290909e+01  1.433751e+01  1.574302e+01   
bowler1_wkts       3.766801e-01  4.309764e-01  4.828872e-01  5.289312e-01   
bowler2_id         6.892765e+05  6.908800e+05  6.908208e+05  6.899128e+05   
bowler2_maidens    1.631784e-02  1.634684e-02  1.644031e-02  1.542878e-02   
bowler2_overs      2.203902e+00  2.200711e+00  2.185418e+00  2.182562e+00   
bowler2_runs       1.811458e+01  1.808173e+01  1.790565e+01  1.790348e+01   
bowler2_wkts       6.002128e-01  5.998579e-01  5.961401e-01  5.905992e-01   
comment_id         8.899079e+04  8.867452e+04  8.781645e+04  8.739668e+04   
innings_id         1.489583e+00  1.489226e+00  1.483226e+00  1.486385e+00   
isBoundary         1.821237e-01  1.882155e-01  1.904439e-01  1.912866e-01   
isNoball           4.032258e-03  4.713805e-03  5.760759e-03  7.488087e-03   
isRetiredHurt      3.360215e-04  0.000000e+00  0.000000e+00  0.000000e+00   
isWide             3.763441e-02  3.804714e-02  3.625890e-02  3.573860e-02   
match_id           1.360081e+06  1.360083e+06  1.360090e+06  1.360078e+06   
over               1.033199e+01  1.032593e+01  1.027516e+01  1.022941e+01   
runs               1.423051e+00  1.453199e+00  1.432057e+00  1.470728e+00   
season             2.023000e+03  2.023000e+03  2.023000e+03  2.023000e+03   
wicket_id          6.060089e+05  6.109063e+05  6.699622e+05  6.251550e+05   
wkt_batsman_balls  1.513953e+01  1.556051e+01  1.697740e+01  1.291034e+01   
wkt_batsman_runs   2.000775e+01  2.003185e+01  2.356497e+01  1.723448e+01   

ball                          5             6  
batsman1_balls     1.402904e+01  1.405028e+01  
batsman1_id        6.134449e+05  6.173807e+05  
batsman1_runs      2.030971e+01  2.038176e+01  
batsman2_balls     1.371690e+01  1.394660e+01  
batsman2_id        6.100765e+05  6.084120e+05  
batsman2_runs      1.934462e+01  1.956311e+01  
bowler1_id         6.946922e+05  6.960561e+05  
bowler1_maidens    1.036986e-02  1.595007e-02  
bowler1_overs      1.747183e+00  2.219626e+00  
bowler1_runs       1.712859e+01  1.852427e+01  
bowler1_wkts       5.741445e-01  6.106103e-01  
bowler2_id         6.905476e+05  6.888244e+05  
bowler2_maidens    1.604668e-02  1.608775e-02  
bowler2_overs      2.175164e+00  2.169031e+00  
bowler2_runs       1.781437e+01  1.776746e+01  
bowler2_wkts       5.875274e-01  5.806216e-01  
comment_id         8.735134e+04  8.655532e+04  
innings_id         1.485655e+00  1.484743e+00  
isBoundary         1.731766e-01  1.851595e-01  
isNoball           5.876253e-03  5.201110e-03  
isRetiredHurt      0.000000e+00  1.040222e-03  
isWide             2.903560e-02  3.259362e-02  
match_id           1.360090e+06  1.360070e+06  
over               1.019253e+01  1.012032e+01  
runs               1.395092e+00  1.448682e+00  
season             2.023000e+03  2.023000e+03  
wicket_id          5.857646e+05  5.920167e+05  
wkt_batsman_balls  1.470748e+01  1.665185e+01  
wkt_batsman_runs   2.050340e+01  2.369630e+01

###### this above pivot table shows avg of every feature for each ball , i.e. same could be done as below

In [14]:
grps=df.groupby(['ball'])
grps.mean().T.sort_index()

C:\Users\Admin\AppData\Local\Temp\ipykernel_5740\1197446882.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  grps.mean().T.sort_index()


ball                          1             2             3             4  \
batsman1_balls     1.431922e+01  1.407407e+01  1.400271e+01  1.380054e+01   
batsman1_id        6.131202e+05  6.119854e+05  6.152186e+05  6.087029e+05   
batsman1_runs      2.019019e+01  2.041987e+01  2.027652e+01  2.000919e+01   
batsman2_balls     1.246808e+01  1.306970e+01  1.327652e+01  1.348502e+01   
batsman2_id        6.073352e+05  6.129574e+05  6.154794e+05  6.128023e+05   
batsman2_runs      1.790894e+01  1.831279e+01  1.868079e+01  1.904187e+01   
bowler1_id         6.966580e+05  6.956873e+05  6.960512e+05  6.949677e+05   
bowler1_maidens    1.108871e-02  1.043771e-02  1.016605e-02  1.055140e-02   
bowler1_overs      1.367036e+00  1.463401e+00  1.559031e+00  1.647754e+00   
bowler1_runs       1.150134e+01  1.290909e+01  1.433751e+01  1.574302e+01   
bowler1_wkts       3.766801e-01  4.309764e-01  4.828872e-01  5.289312e-01   
bowler2_id         6.892765e+05  6.908800e+05  6.908208e+05  6.899128e+05   
bowler2_maidens    1.631784e-02  1.634684e-02  1.644031e-02  1.542878e-02   
bowler2_overs      2.203902e+00  2.200711e+00  2.185418e+00  2.182562e+00   
bowler2_runs       1.811458e+01  1.808173e+01  1.790565e+01  1.790348e+01   
bowler2_wkts       6.002128e-01  5.998579e-01  5.961401e-01  5.905992e-01   
comment_id         8.899079e+04  8.867452e+04  8.781645e+04  8.739668e+04   
innings_id         1.489583e+00  1.489226e+00  1.483226e+00  1.486385e+00   
isBoundary         1.821237e-01  1.882155e-01  1.904439e-01  1.912866e-01   
isNoball           4.032258e-03  4.713805e-03  5.760759e-03  7.488087e-03   
isRetiredHurt      3.360215e-04  0.000000e+00  0.000000e+00  0.000000e+00   
isWide             3.763441e-02  3.804714e-02  3.625890e-02  3.573860e-02   
match_id           1.360081e+06  1.360083e+06  1.360090e+06  1.360078e+06   
over               1.033199e+01  1.032593e+01  1.027516e+01  1.022941e+01   
runs               1.423051e+00  1.453199e+00  1.432057e+00  1.470728e+00   
season             2.023000e+03  2.023000e+03  2.023000e+03  2.023000e+03   
wicket_id          6.060089e+05  6.109063e+05  6.699622e+05  6.251550e+05   
wkt_batsman_balls  1.513953e+01  1.556051e+01  1.697740e+01  1.291034e+01   
wkt_batsman_runs   2.000775e+01  2.003185e+01  2.356497e+01  1.723448e+01   

ball                          5             6  
batsman1_balls     1.402904e+01  1.405028e+01  
batsman1_id        6.134449e+05  6.173807e+05  
batsman1_runs      2.030971e+01  2.038176e+01  
batsman2_balls     1.371690e+01  1.394660e+01  
batsman2_id        6.100765e+05  6.084120e+05  
batsman2_runs      1.934462e+01  1.956311e+01  
bowler1_id         6.946922e+05  6.960561e+05  
bowler1_maidens    1.036986e-02  1.595007e-02  
bowler1_overs      1.747183e+00  2.219626e+00  
bowler1_runs       1.712859e+01  1.852427e+01  
bowler1_wkts       5.741445e-01  6.106103e-01  
bowler2_id         6.905476e+05  6.888244e+05  
bowler2_maidens    1.604668e-02  1.608775e-02  
bowler2_overs      2.175164e+00  2.169031e+00  
bowler2_runs       1.781437e+01  1.776746e+01  
bowler2_wkts       5.875274e-01  5.806216e-01  
comment_id         8.735134e+04  8.655532e+04  
innings_id         1.485655e+00  1.484743e+00  
isBoundary         1.731766e-01  1.851595e-01  
isNoball           5.876253e-03  5.201110e-03  
isRetiredHurt      0.000000e+00  1.040222e-03  
isWide             2.903560e-02  3.259362e-02  
match_id           1.360090e+06  1.360070e+06  
over               1.019253e+01  1.012032e+01  
runs               1.395092e+00  1.448682e+00  
season             2.023000e+03  2.023000e+03  
wicket_id          5.857646e+05  5.920167e+05  
wkt_batsman_balls  1.470748e+01  1.665185e+01  
wkt_batsman_runs   2.050340e+01  2.369630e+01